File: fushi/test/pages/popup_glossary_export_parity_test.js
// BUG-1062 / BUG-1061 behavior test: the Anki mining payload built by popup.js
// must match upstream Yomitan's exported glossary, in two respects the user hit:
//
// BUG-1061 — `{glossary}` label had a self-invented ordinal. Yomitan's
//   `glossary-single` anki template emits `(definitionTags…, dictionaryAlias)`
//   with NO number, and this repo's own constructSingleGlossaryHtml
//   (`{glossary-first}` / `{single-glossary-*}`) already agreed. Only
//   constructGlossaryHtml prefixed an index, so cards read "(1, 词典名)".
//
// BUG-1062 — exported definition images were pinned to physical pixels. Yomitan's
//   structured-content-generator always writes `width: {usedWidth}em` on the
//   container and lets CSS decide what 1em is: in its own popup a stylesheet
//   squashes it to ~1px, but an Anki card has no such stylesheet, so `em`
//   resolves against the card font size. popup.js instead exported
//   `width: {usedWidth}px` plus an inline `font-size: 1px` (which, being inline,
//   also overrode any note-type CSS) — cards ended up a whole font-size factor
//   smaller than Yomitan's. Export now keeps the `em` semantics; the popup path
//   keeps px (its own CSS is what makes px correct there).
//
// This EXECUTES the real popup.js against a minimal fake DOM. Reverting either
// fix turns this red.
//
// Run: node fushi/test/pages/popup_glossary_export_parity_test.js
// (also driven from popup_glossary_export_parity_test.dart inside `flutter test`).

const assert = require('assert');
const fs = require('fs');
const path = require('path');
const vm = require('vm');

const popupPath = path.resolve(__dirname, '../../assets/popup/popup.js');
// popup.html loads dict-media.js before popup.js; the export path calls into it
// (normalizeDictMediaPath / rewriteDictionaryMediaPath), so load the real thing
// rather than stubbing the media path rules.
const dictMediaPath = path.resolve(__dirname, '../../assets/popup/dict-media.js');
const source = fs.readFileSync(dictMediaPath, 'utf8') + '\n' +
  fs.readFileSync(popupPath, 'utf8');

function makeElement(tag) {
  return {
    tagName: (tag || 'div').toUpperCase(),
    className: '',
    id: '',
    textContent: '',
    innerHTML: '',
    // Real CSSStyleDeclaration semantics that popup.js relies on: individual
    // property writes plus `cssText +=` accumulation.
    style: { cssText: '' },
    dataset: {},
    children: [],
    attributes: [],
    classList: {
      _set: new Set(),
      add(name) { this._set.add(name); },
      remove(name) { this._set.delete(name); },
      contains(name) { return this._set.has(name); },
    },
    appendChild(child) { this.children.push(child); return child; },
    append(...nodes) { this.children.push(...nodes); },
    setAttribute() {},
    getAttribute() { return null; },
    hasAttribute() { return false; },
    removeAttribute() {},
    addEventListener() {},
    querySelectorAll() { return []; },
    querySelector() { return null; },
    closest() { return null; },
    get firstChild() { return this.children.length ? this.children[0] : null; },
  };
}

function makeSandbox() {
  const documentObj = {
    documentElement: { style: {}, classList: makeElement().classList },
    head: { appendChild() {} },
    body: makeElement('body'),
    getElementById() { return null; },
    querySelector() { return null; },
    querySelectorAll() { return []; },
    createElement(tag) { return makeElement(tag); },
    createTextNode(text) { const n = makeElement('#text'); n.textContent = text; return n; },
    addEventListener() {},
  };

  const windowObj = {
    audioSources: [],
    needsAudio: false,
    lookupEntries: [],
    dictionaryStyles: {},
    hiddenDictionaryNames: [],
    collapsedDictionaryNames: [],
    compactGlossariesAnki: false,
    // Mining payload path: dictionary media is embedded, so exported images are
    // <img src="fushi_dict_N.ext"> and go through applyImageStyles.
    embedMedia: true,
    devicePixelRatio: 2,
    innerWidth: 400,
    flutter_inappwebview: { callHandler() { return Promise.resolve(false); } },
    getSelection() { return { toString() { return ''; } }; },
  };
  documentObj.defaultView = windowObj;

  const sandbox = {
    Node: { TEXT_NODE: 3, ELEMENT_NODE: 1 },
    Date, Math, URL, JSON, RegExp, Set, Map, Object, Array, console,
    performance: { now() { return 0; } },
    setTimeout, clearTimeout,
    DOMParser: class { parseFromString() { return { body: makeElement('body'), querySelectorAll() { return []; } }; } },
    document: documentObj,
    window: windowObj,
    getComputedStyle() { return {}; },
  };
  sandbox.globalThis = sandbox;
  return sandbox;
}

function loadPopup(entry) {
  const sandbox = makeSandbox();
  vm.createContext(sandbox);
  const exported = source + `
    ;window.lookupEntries = [${JSON.stringify(entry)}];
    // buildMinePayload normally opens this registry before rendering fields.
    ;currentDictionaryMedia = new Map();
    ;window.__test = {
      multi: function() { return constructGlossaryHtml(0); },
      single: function() { return constructSingleGlossaryHtml(0); },
      image: function(data, exporting) { return createDefinitionImage(data, 'Dict', exporting); },
      structured: function(parent, node, exporting) {
        return renderStructuredContent(parent, node, null, 'Dict', exporting);
      },
    };
  `;
  vm.runInContext(exported, sandbox, { filename: 'popup.js' });
  return sandbox;
}

function gloss(dictionary, text) {
  return { dictionary: dictionary, content: text, definitionTags: '', termTags: '' };
}

function labelsOf(html) {
  const out = [];
  const re = /<i>([^<]*)<\/i>/g;
  let m;
  while ((m = re.exec(html)) !== null) { out.push(m[1]); }
  return out;
}

// The container span is the first child of the returned .gloss-image-link node.
function containerOf(node) {
  assert.ok(node && node.children.length > 0, 'image node must have a container child');
  return node.children[0];
}

(function run() {
  const entry = {
    expression: '猫', reading: 'ねこ',
    glossaries: [gloss('JMdict', 'cat'), gloss('JMdict', 'kitty'), gloss('Daijirin', 'ねこ科の動物')],
    frequencies: [], pitches: [],
  };

  // BUG-1061 (1/2): no ordinal anywhere in the {glossary} labels — the first
  // entry of a dictionary carries just the dictionary name, subsequent entries
  // of the same dictionary carry nothing (tags would go here when present).
  {
    const sb = loadPopup(entry);
    const multi = sb.window.__test.multi();
    assert.deepStrictEqual(labelsOf(multi), ['(JMdict)', '', '(Daijirin)'],
      'the {glossary} labels must be Yomitan-shaped and ordinal-free; got ' + multi);
    assert.ok(!/\(\d+[,)]/.test(multi),
      'no "(1," / "(1)" ordinal may appear in the exported glossary; got ' + multi);
  }

  // BUG-1061 (2/2): the two mining payload builders must agree on the label
  // format — they render the same dictionary for the same card.
  {
    const sb = loadPopup(entry);
    const multi = sb.window.__test.multi();
    const single = sb.window.__test.single();
    for (const dict of ['JMdict', 'Daijirin']) {
      const first = labelsOf(single[dict])[0];
      assert.ok(labelsOf(multi).includes(first),
        `{glossary} must use the same label as {glossary-first} for ${dict}; ` +
        `single=${first} multi=${JSON.stringify(labelsOf(multi))}`);
    }
  }

  // BUG-1062 (1/3): an exported image keeps Yomitan's em sizing, and does not
  // pin an inline 1px font-size (which would also outrank note-type CSS).
  {
    const sb = loadPopup(entry);
    const node = sb.window.__test.image({ path: 'pic.png', width: 10, height: 5 }, true);
    const container = containerOf(node);
    assert.strictEqual(container.style.width, '10em',
      'exported image container must size in em like Yomitan; got ' + container.style.width);
    assert.ok(/font-size:1em/.test(container.style.cssText),
      'exported image container must inherit the card font size; got ' + container.style.cssText);
    assert.ok(!/font-size:1px/.test(container.style.cssText),
      'exported image container must not pin font-size:1px (BUG-1062); got ' + container.style.cssText);
    assert.ok(/max-width:100%/.test(container.style.cssText),
      'exported image must still be capped at the card width');
  }

  // BUG-1062 (2/3): the popup path is unchanged — px there is correct because
  // the popup ships the Yomitan stylesheet.
  {
    const sb = loadPopup(entry);
    const node = sb.window.__test.image({ path: 'pic.png', width: 10, height: 5 }, false);
    const container = containerOf(node);
    assert.strictEqual(container.style.width, '10px',
      'popup image container must keep px sizing; got ' + container.style.width);
  }

  // BUG-1062 (3/3): dictionaries that declare em units are untouched by the fix
  // (they were already em on both paths).
  {
    const sb = loadPopup(entry);
    const node = sb.window.__test.image(
      { path: 'gaiji.svg', width: 2, height: 2, sizeUnits: 'em' }, true);
    const container = containerOf(node);
    assert.strictEqual(container.style.width, '2em',
      'sizeUnits:em images stay em on export; got ' + container.style.width);
  }

  // BUG-1676 (1/3): a dictionary that declares NO size must not be exported at
  // the `width = 100` fallback — that number is neither the image's size nor its
  // aspect ratio, and as `100em` it resolves to ~100 x card-font-size (2000px on
  // a 775px-wide card). The <img> itself becomes the layout box instead.
  {
    const sb = loadPopup(entry);
    const node = sb.window.__test.image({ path: 'pic.png' }, true);
    const container = containerOf(node);
    assert.strictEqual(container.style.width, 'auto',
      'an image with no declared size must be laid out by the <img>; got ' + container.style.width);
    assert.ok(!/100em/.test(container.style.cssText + container.style.width),
      'the 100 fallback must never be exported as 100em (BUG-1676); got ' + container.style.cssText);
    assert.strictEqual(node.dataset.hasAspectRatio, 'false',
      'no declared size means no invented 1:1 aspect ratio; got ' + node.dataset.hasAspectRatio);
    const img = container.children.find(c => c.tagName === 'IMG');
    assert.ok(img, 'exported node must contain an <img>');
    assert.ok(img.width === undefined && img.height === undefined,
      'the 100x100 fallback must not be written as width/height attributes; got ' +
      img.width + 'x' + img.height);
    assert.ok(/position:static/.test(img.style.cssText) && /max-width:100%/.test(img.style.cssText),
      'the natural-size <img> must be in flow and capped at the card width; got ' + img.style.cssText);
    assert.ok(!/position:absolute/.test(img.style.cssText),
      'an absolutely positioned <img> would collapse the auto-width container to 0; got ' +
      img.style.cssText);
  }

  // BUG-1676 (2/3): declared sizes are untouched — the BUG-1062 em semantics
  // still apply, and the aspect-ratio sizer still drives the layout.
  {
    const sb = loadPopup(entry);
    const node = sb.window.__test.image({ path: 'pic.png', width: 10, height: 5 }, true);
    const container = containerOf(node);
    assert.strictEqual(container.style.width, '10em',
      'declared-size images keep Yomitan em sizing; got ' + container.style.width);
    assert.strictEqual(node.dataset.hasAspectRatio, 'true',
      'declared-size images keep their aspect-ratio sizer');
    const img = container.children.find(c => c.tagName === 'IMG');
    assert.ok(/position:absolute/.test(img.style.cssText),
      'declared-size images keep the sizer + absolute <img> layout; got ' + img.style.cssText);
  }

  // BUG-1676 (3/3): exported structured-content tables carry the overflow guard
  // inline. The popup gets it from popup.css; an Anki card has no stylesheet, and
  // `table-layout:auto` ignores percentage max-width on cells — without this the
  // whole card is dragged off the right edge of the screen by one wide table.
  {
    const sb = loadPopup(entry);
    const parent = sb.document.createElement('div');
    sb.window.__test.structured(parent, { tag: 'table', content: 'x' }, true);
    const container = parent.children[0];
    assert.ok(container && container.classList.contains('gloss-sc-table-container'),
      'a structured-content table must be wrapped in .gloss-sc-table-container');
    assert.ok(/overflow-x:auto/.test(container.style.cssText),
      'exported table container must inline the overflow guard (BUG-1676); got ' +
      container.style.cssText);
    assert.ok(/max-width:100%/.test(container.style.cssText),
      'exported table container must be capped at the card width; got ' + container.style.cssText);

    const popupParent = sb.document.createElement('div');
    sb.window.__test.structured(popupParent, { tag: 'table', content: 'x' }, false);
    assert.strictEqual(popupParent.children[0].style.cssText, '',
      'the popup path must stay CSS-driven, not inline-styled');
  }

  console.log('popup_glossary_export_parity_test.js: all assertions passed');
})();

File: fushi/test/pages/popup_glossary_export_parity_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// BUG-1062 / BUG-1061：制卡（Anki mining）产出的 glossary 字段必须与上游 Yomitan
/// 导出的卡片一致，用户在两处看到了差异。
///
/// **BUG-1061 — `{glossary}` 词典名前多一个自造序号。**
/// 上游 Yomitan 的 anki 模板 `glossary-single` 标签是
/// `(definitionTags…, dictionaryAlias)`，**没有序号**；本仓 `constructSingleGlossaryHtml`
/// （`{glossary-first}` / `{single-glossary-*}` 用）也一直是这个格式。只有
/// `constructGlossaryHtml`（`{glossary}` 用）自增了一个 `index` 塞进标签，卡片上就成了
/// 「(1, 词典名)」。修复：删掉 index，两个 builder 标签格式统一。
///
/// **BUG-1062 — 导出图片被钉成物理像素，比 Yomitan 卡片小一个字号的倍数。**
/// 上游 `structured-content-generator.js` 导出时容器**永远**写 `width: {usedWidth}em`，
/// 1em 具体多大交给 CSS：Yomitan 弹窗自带
/// `.gloss-image-container{font-size:calc(1em/var(--font-size-no-units))}`（≈1px），
/// 而 Anki 卡片上没有这份 CSS，em 按卡片正文字号解析——这就是 Yomitan 卡片图「大」的原因。
/// popup.js 导出时却把它折算成 `width:{usedWidth}px` 且内联 `font-size:1px`（内联优先级
/// 最高，note type 的 CSS 也压不动），于是卡片图小了约一个字号的倍数。修复：导出保留 em
/// 语义、不再钉 font-size；弹窗路径维持 px（弹窗自己带那份 CSS，px 在那里才是对的）。
///
/// 三层守护：
/// ① 行为级——用 Node 真执行 popup.js 的 `constructGlossaryHtml` /
///    `constructSingleGlossaryHtml` / `createDefinitionImage`（见同名 .js）。无 node 时 skip。
/// ② 源码级——静态断言导出分支用 em、不再内联 `font-size:1px`，标签里没有序号。
/// ③ 三镜像——app 内弹窗 / 扩展 vendor 两份镜像与主文件在这些点上必须一致
///    （`tools/browser-extension/vendor/popup.js` 由 browser_extension_popup_parity_guard
///    另行守全量一致性，这里只保证本修复不漏改镜像）。
void main() {
  test(
    'mining glossary matches Yomitan (no ordinal, em-sized images) '
    '(executes popup.js via node)',
    () async {
      final String? nodeExe = _resolveNode();
      if (nodeExe == null) {
        markTestSkipped(
            'node not found on PATH; skipping JS behavior execution');
        return;
      }

      final File jsTest =
          File('test/pages/popup_glossary_export_parity_test.js');
      expect(
        jsTest.existsSync(),
        isTrue,
        reason: 'behavior harness ${jsTest.path} must exist',
      );

      final ProcessResult result = await Process.run(
        nodeExe,
        <String>[jsTest.path],
        workingDirectory: Directory.current.path,
      );

      expect(
        result.exitCode,
        0,
        reason: 'glossary export parity JS behavior test failed.\n'
            'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
      );
      expect(
        result.stdout.toString(),
        contains('all assertions passed'),
        reason: 'behavior harness must reach its success marker',
      );
    },
  );

  test('popup.js mirrors keep the Yomitan-shaped mining glossary', () {
    for (final String relative in _mirrors) {
      final File file = File(relative);
      expect(file.existsSync(), isTrue, reason: '$relative must exist');
      final String js = file.readAsStringSync();

      // BUG-1061: neither mining builder may prefix an ordinal to the label.
      expect(
        RegExp(r'label = tags \? `\(\$\{index\}').hasMatch(js),
        isFalse,
        reason: '$relative: the {glossary} label must not carry an ordinal '
            '(BUG-1061)',
      );
      expect(
        js.contains(r'let index = 0;'),
        isFalse,
        reason: '$relative: the self-invented glossary ordinal counter must be '
            'gone (BUG-1061)',
      );

      // BUG-1062: the export branch sizes in em; the popup branch keeps px.
      expect(
        js.contains(
            r'imageContainer.style.width = exporting ? `${usedWidth}em` : `${usedWidth}px`;'),
        isTrue,
        reason: '$relative: exported definition images must keep Yomitan em '
            'sizing while the popup keeps px (BUG-1062)',
      );
      expect(
        js.contains('font-size:1px'),
        isFalse,
        reason:
            '$relative: the exported image container must not pin an inline '
            'font-size:1px — it outranks note-type CSS and shrinks the card '
            'image (BUG-1062)',
      );
    }
  });
}

const List<String> _mirrors = <String>[
  'assets/popup/popup.js',
  'assets/browser_extension/vendor/popup.js',
  '../tools/browser-extension/vendor/popup.js',
];

/// Resolve a usable `node` executable, returning null when none is on PATH.
String? _resolveNode() {
  final List<String> candidates =
      Platform.isWindows ? <String>['node.exe', 'node'] : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) {
        return name;
      }
    } on ProcessException {
      // Not found; try next candidate.
    }
  }
  return null;
}
